flink-runtime/src/test/java/org/apache/flink/runtime/io/network/buffer/NetworkBufferPoolTest.java

```diff
@@ -518,4 +528,176 @@ public void go() throws Exception {
 			globalPool.destroy();
 		}
 	}
+
+	/**
+	 * Tests {@link NetworkBufferPool#isAvailable()}, verifying that the buffer availability is
+	 * correctly maintained and the future callback is correctly processed.
+	 */
+	@Test
+	public void testBufferAvailabilityAndFutureCallback() throws Exception {
+		final int numBuffers = 10;
+		final int numberOfSegmentsToRequest = 5;
+		final Duration requestSegmentsTimeout = Duration.ofSeconds(10L);
+
+		final NetworkBufferPool globalPool = new NetworkBufferPool(
+				numBuffers,
+				128,
+				numberOfSegmentsToRequest,
+				requestSegmentsTimeout);
+
+		try {
+			assertTrue(globalPool.isAvailable().isDone());
+
+			List<MemorySegment> segments = new ArrayList<>(numberOfSegmentsToRequest);
+			for (int i = 0; i < numberOfSegmentsToRequest; ++i) {
+				MemorySegment segment = globalPool.requestMemorySegment();
+				assertNotNull(segment);
+				segments.add(segment);
+				assertTrue(globalPool.isAvailable().isDone());
+			}
+
+			final List<MemorySegment> exclusiveSegments = globalPool.requestMemorySegments();
+			assertEquals(numberOfSegmentsToRequest, exclusiveSegments.size());
+
+			assertFalse(globalPool.isAvailable().isDone());
+			assertEquals(0, globalPool.getNumberOfAvailableMemorySegments());
+
+			final AtomicBoolean completeFlag = new AtomicBoolean(false);
+			CompletableFuture<?> availableFuture = globalPool.isAvailable();
+			availableFuture.whenComplete((ignored, throwable) -> completeFlag.set(true));
+
+			// recycle one buffer
+			globalPool.recycle(segments.get(0));
+			assertTrue(completeFlag.get());
+			assertTrue(availableFuture.isDone());
+			assertTrue(globalPool.isAvailable().isDone());
+			assertEquals(1, globalPool.getNumberOfAvailableMemorySegments());
```

Comment:
ditto: we could extract a separate test for verifying the `globalPool.requestMemorySegments()`.